server/src/main/java/org/elasticsearch/action/bulk/BulkResponse.java

```diff
@@ -161,9 +175,14 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             builder.field(INGEST_TOOK, ingestTookInMillis);
         }
         builder.field(ERRORS, hasFailures());
+        boolean noItems = noItemsOnSuccess != null && noItemsOnSuccess;
+        boolean noFailuresAnNoItemsRequested = noItems && this.hasFailures() == false;
+        builder.field(ITEMS_OMITTED, noFailuresAnNoItemsRequested);
         builder.startArray(ITEMS);
-        for (BulkItemResponse item : this) {
-            item.toXContent(builder, params);
+        if (noFailuresAnNoItemsRequested == false) {
```

Comment:
Would be great if the decision on whether to include the items can be made earlier.
Maybe in `TransportBulkAction` in the `finishHim()` method around line 541?
This would then save creating bulk item response array. 

Would be great if we could make this decision even earlier, so that no shard level response items
are collected. Then in the case that there are failures then we could serialize the successful items
as null elements? But then I'm afraid that we break the bulk response format.

Reply:
Yeah I could do the first, that would definitely help 👍 

Not sure the second warrants the additional complexity? The goal of this PR is to shave off serialization time on the consumer. Perhaps something to consider once this lands?

Reply:
@martijnvg revisiting this I think it's best not to introduce the complexity in the transport layer and only expose this on the REST layer. 

It would introduce a set of new problems 
* shards would need to send back how many items they would have returned
* reducing all shard information becomes more complex to expand the `empty` shard responses
* We need to find a new representation for successful bulk operations in case of failures.

The PR as it stands does not preclude taking this on as follow up work but is useful as is already.

Keen to hear your thoughts!

Reply:
@Mpdreamz I agree, this PR is useful in its current form. 

However I think it makes sense to not include the items in the bulk response if there are no failures and no_items_on_success is set. This can simply be done by iterating over the response items and see if there is an error and then have an additional constructor of bulk responses which accepts no response items and `noFailuresAnNoItemsRequested` (or sets that always to true).